// @shopify/shopify-app-remix/build/ts/server/authenticate/admin/helpers/handle-client-error.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.handleClientErrorFactory = void 0;
var tslib_1 = require("tslib");
var shopify_api_1 = require("@shopify/shopify-api");
var redirect_to_auth_page_1 = require("./redirect-to-auth-page");
function handleClientErrorFactory(_a) {
    var request = _a.request;
    return function handleClientError(_a) {
        var error = _a.error, params = _a.params, session = _a.session;
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            return tslib_1.__generator(this, function (_b) {
                switch (_b.label) {
                    case 0:
                        if (error instanceof shopify_api_1.HttpResponseError !== true) {
                            params.logger.debug("Got a response error from the API: ".concat(error.message));
                            throw error;
                        }
                        params.logger.debug("Got an HTTP response error from the API: ".concat(error.message), {
                            code: error.response.code,
                            statusText: error.response.statusText,
                            body: JSON.stringify(error.response.body),
                        });
                        if (!(error.response.code === 401)) return [3 /*break*/, 2];
                        return [4 /*yield*/, (0, redirect_to_auth_page_1.redirectToAuthPage)(params, request, session.shop)];
                    case 1: throw _b.sent();
                    case 2: 
                    // forward a minimal copy of the upstream HTTP response instead of an Error:
                    throw new Response(JSON.stringify(error.response.body), {
                        status: error.response.code,
                        headers: {
                            'Content-Type': error.response.headers['Content-Type'],
                        },
                    });
                }
            });
        });
    };
}
exports.handleClientErrorFactory = handleClientErrorFactory;
