// @shopify/shopify-app-remix/build/ts/server/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.shopifyApp = exports.boundary = exports.AppDistribution = exports.LoginErrorType = exports.ApiVersion = exports.BillingInterval = exports.DeliveryMethod = exports.LogSeverity = exports.LATEST_API_VERSION = void 0;
require("@shopify/shopify-api/adapters/web-api");
var runtime_1 = require("@shopify/shopify-api/runtime");
(0, runtime_1.setAbstractRuntimeString)(function () {
    return "Remix";
});
var shopify_api_1 = require("@shopify/shopify-api");
Object.defineProperty(exports, "LATEST_API_VERSION", { enumerable: true, get: function () { return shopify_api_1.LATEST_API_VERSION; } });
Object.defineProperty(exports, "LogSeverity", { enumerable: true, get: function () { return shopify_api_1.LogSeverity; } });
Object.defineProperty(exports, "DeliveryMethod", { enumerable: true, get: function () { return shopify_api_1.DeliveryMethod; } });
Object.defineProperty(exports, "BillingInterval", { enumerable: true, get: function () { return shopify_api_1.BillingInterval; } });
Object.defineProperty(exports, "ApiVersion", { enumerable: true, get: function () { return shopify_api_1.ApiVersion; } });
var types_1 = require("./types");
Object.defineProperty(exports, "LoginErrorType", { enumerable: true, get: function () { return types_1.LoginErrorType; } });
Object.defineProperty(exports, "AppDistribution", { enumerable: true, get: function () { return types_1.AppDistribution; } });
var boundary_1 = require("./boundary");
Object.defineProperty(exports, "boundary", { enumerable: true, get: function () { return boundary_1.boundary; } });
var shopify_app_1 = require("./shopify-app");
Object.defineProperty(exports, "shopifyApp", { enumerable: true, get: function () { return shopify_app_1.shopifyApp; } });
