// @shopify/shopify-api/lib/auth/oauth/token-exchange.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.tokenExchange = exports.RequestedTokenType = void 0;
const tslib_1 = require("tslib");
const decode_session_token_1 = require("../../session/decode-session-token");
const shop_validator_1 = require("../../utils/shop-validator");
const types_1 = require("../../clients/http_client/types");
const http_client_1 = require("../../clients/http_client/http_client");
const create_session_1 = require("./create-session");
var RequestedTokenType;
(function (RequestedTokenType) {
    RequestedTokenType["OnlineAccessToken"] = "urn:shopify:params:oauth:token-type:online-access-token";
    RequestedTokenType["OfflineAccessToken"] = "urn:shopify:params:oauth:token-type:offline-access-token";
})(RequestedTokenType = exports.RequestedTokenType || (exports.RequestedTokenType = {}));
const TokenExchangeGrantType = 'urn:ietf:params:oauth:grant-type:token-exchange';
const IdTokenType = 'urn:ietf:params:oauth:token-type:id_token';
function tokenExchange(config) {
    return ({ shop, sessionToken, requestedTokenType, }) => tslib_1.__awaiter(this, void 0, void 0, function* () {
        yield (0, decode_session_token_1.decodeSessionToken)(config)(sessionToken);
        const body = {
            client_id: config.apiKey,
            client_secret: config.apiSecretKey,
            grant_type: TokenExchangeGrantType,
            subject_token: sessionToken,
            subject_token_type: IdTokenType,
            requested_token_type: requestedTokenType,
        };
        const postParams = {
            path: '/admin/oauth/access_token',
            type: types_1.DataType.JSON,
            data: body,
            extraHeaders: {
                Accept: types_1.DataType.JSON,
            },
        };
        const cleanShop = (0, shop_validator_1.sanitizeShop)(config)(shop, true);
        const HttpClient = (0, http_client_1.httpClientClass)(config);
        const client = new HttpClient({ domain: cleanShop });
        const postResponse = yield client.post(postParams);
        return {
            session: (0, create_session_1.createSession)({
                accessTokenResponse: (0, create_session_1.accessTokenResponse)(postResponse),
                shop: cleanShop,
                // We need to keep this as an empty string as our template DB schemas have this required
                state: '',
                config,
            }),
        };
    });
}
exports.tokenExchange = tokenExchange;
