// @shopify/shopify-app-remix/build/cjs/server/index.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

require('@shopify/shopify-api/adapters/web-api');
var runtime = require('@shopify/shopify-api/runtime');
var shopifyApi = require('@shopify/shopify-api');
var types = require('./types.js');
var index = require('./boundary/index.js');
var shopifyApp = require('./shopify-app.js');

runtime.setAbstractRuntimeString(() => {
  return `Remix`;
});

Object.defineProperty(exports, 'ApiVersion', {
  enumerable: true,
  get: function () { return shopifyApi.ApiVersion; }
});
Object.defineProperty(exports, 'BillingInterval', {
  enumerable: true,
  get: function () { return shopifyApi.BillingInterval; }
});
Object.defineProperty(exports, 'DeliveryMethod', {
  enumerable: true,
  get: function () { return shopifyApi.DeliveryMethod; }
});
Object.defineProperty(exports, 'LATEST_API_VERSION', {
  enumerable: true,
  get: function () { return shopifyApi.LATEST_API_VERSION; }
});
Object.defineProperty(exports, 'LogSeverity', {
  enumerable: true,
  get: function () { return shopifyApi.LogSeverity; }
});
exports.AppDistribution = types.AppDistribution;
exports.LoginErrorType = types.LoginErrorType;
exports.boundary = index.boundary;
exports.shopifyApp = shopifyApp.shopifyApp;
