// @shopify/shopify-app-remix/build/ts/server/clients/storefront/factory.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.storefrontClientFactory = void 0;
var tslib_1 = require("tslib");
var runtime_1 = require("@shopify/shopify-api/runtime");
function storefrontClientFactory(_a) {
    var _this = this;
    var params = _a.params, session = _a.session;
    var api = params.api;
    return {
        graphql: function (query, options) {
            if (options === void 0) { options = {}; }
            return tslib_1.__awaiter(_this, void 0, void 0, function () {
                var client, apiResponse;
                return tslib_1.__generator(this, function (_a) {
                    switch (_a.label) {
                        case 0:
                            client = new api.clients.Storefront({
                                session: session,
                                apiVersion: options.apiVersion,
                            });
                            return [4 /*yield*/, client.query({
                                    data: { query: query, variables: options === null || options === void 0 ? void 0 : options.variables },
                                    tries: options.tries,
                                    extraHeaders: options.headers,
                                })];
                        case 1:
                            apiResponse = _a.sent();
                            return [2 /*return*/, new Response(JSON.stringify(apiResponse.body), {
                                    headers: (0, runtime_1.flatHeaders)(apiResponse.headers),
                                })];
                    }
                });
            });
        },
    };
}
exports.storefrontClientFactory = storefrontClientFactory;
